{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5a56362b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import time\n",
    "import re\n",
    "import easygui as eg\n",
    "from selenium import webdriver\n",
    "from selenium.webdriver.common.by import By\n",
    "from selenium.webdriver.support.ui import WebDriverWait\n",
    "from selenium.webdriver.support import expected_conditions as EC\n",
    "from selenium.webdriver.firefox.options import Options\n",
    "from selenium.common.exceptions import TimeoutException\n",
    "from pdfreader import PDFDocument, SimplePDFViewer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8a0a9025",
   "metadata": {},
   "outputs": [],
   "source": [
    "# =========================\n",
    "#   INTERACTIVE WINDOW\n",
    "# =========================\n",
    "# Interactive window for species choice \n",
    "species_str = [\n",
    "    f\"{scientific} : {short}\"\n",
    "    for short, scientific in [\n",
    "        ('human', 'Homo sapiens'),\n",
    "        ('fly', 'Drosophila melanogaster'),\n",
    "        ('arabidopsis', 'Arabidopsis thaliana'),\n",
    "        ('brugia', 'Brugia malayi'),\n",
    "        ('aedes', 'Aedes aegypti'),\n",
    "        ('tribolium', 'Tribolium castaneum'),\n",
    "        ('schistosoma', 'Schistosoma mansoni'),\n",
    "        ('tetrahymena', 'Tetrahymena thermophila'),\n",
    "        ('galdieria', 'Galdieria sulphuraria'),\n",
    "        ('maize', 'Zea mays'),\n",
    "        ('toxoplasma', 'Toxoplasma gondii'),\n",
    "        ('caenorhabditis', 'Caenorhabditis elegans'),\n",
    "        ('(elegans)', 'Caenorhabditis elegans'),\n",
    "        ('aspergillus_fumigatus', 'Aspergillus fumigatus'),\n",
    "        ('aspergillus_nidulans', 'Aspergillus nidulans'),\n",
    "        ('(anidulans)', 'Aspergillus nidulans'),\n",
    "        ('aspergillus_oryzae', 'Aspergillus oryzae'),\n",
    "        ('aspergillus_terreus', 'Aspergillus terreus'),\n",
    "        ('botrytis_cinerea', 'Botrytis cinerea'),\n",
    "        ('candida_albicans', 'Candida albicans'),\n",
    "        ('candida_guilliermondii', 'Candida guilliermondii'),\n",
    "        ('candida_tropicalis', 'Candida tropicalis'),\n",
    "        ('chaetomium_globosum', 'Chaetomium globosum'),\n",
    "        ('coccidioides_immitis', 'Coccidioides immitis'),\n",
    "        ('coprinus', 'Coprinus cinereus'),\n",
    "        ('coprinus_cinereus', 'Coprinus cinereus'),\n",
    "        ('coyote_tobacco', 'Nicotiana attenuata'),\n",
    "        ('cryptococcus_neoformans_gattii', 'Cryptococcus neoformans gattii'),\n",
    "        ('cryptococcus_neoformans_neoformans_B', 'Cryptococcus neoformans neoformans'),\n",
    "        ('cryptococcus_neoformans_neoformans_JEC21', 'Cryptococcus neoformans neoformans'),\n",
    "        ('(cryptococcus)', 'Cryptococcus neoformans'),\n",
    "        ('debaryomyces_hansenii', 'Debaryomyces hansenii'),\n",
    "        ('encephalitozoon_cuniculi_GB', 'Encephalitozoon cuniculi'),\n",
    "        ('eremothecium_gossypii', 'Eremothecium gossypii'),\n",
    "        ('fusarium_graminearum', 'Fusarium graminearum'),\n",
    "        ('(fusarium)', 'Fusarium graminearum'),\n",
    "        ('histoplasma_capsulatum', 'Histoplasma capsulatum'),\n",
    "        ('(histoplasma)', 'Histoplasma capsulatum'),\n",
    "        ('kluyveromyces_lactis', 'Kluyveromyces lactis'),\n",
    "        ('laccaria_bicolor', 'Laccaria bicolor'),\n",
    "        ('lamprey', 'Petromyzon marinus'),\n",
    "        ('leishmania_tarentolae', 'Leishmania tarentolae'),\n",
    "        ('lodderomyces_elongisporus', 'Lodderomyces elongisporus'),\n",
    "        ('magnaporthe_grisea', 'Magnaporthe grisea'),\n",
    "        ('neurospora_crassa', 'Neurospora crassa'),\n",
    "        ('(neurospora)', 'Neurospora crassa'),\n",
    "        ('phanerochaete_chrysosporium', 'Phanerochaete chrysosporium'),\n",
    "        ('(pchrysosporium)', 'Phanerochaete chrysosporium'),\n",
    "        ('pichia_stipitis', 'Pichia stipitis'),\n",
    "        ('rhizopus_oryzae', 'Rhizopus oryzae'),\n",
    "        ('saccharomyces_cerevisiae_S288C', 'Saccharomyces cerevisiae'),\n",
    "        ('saccharomyces_cerevisiae_rm11-1a_1', 'Saccharomyces cerevisiae'),\n",
    "        ('(saccharomyces)', 'Saccharomyces cerevisiae'),\n",
    "        ('schizosaccharomyces_pombe', 'Schizosaccharomyces pombe'),\n",
    "        ('thermoanaerobacter_tengcongensis', 'Thermoanaerobacter tengcongensis'),\n",
    "        ('trichinella', 'Trichinella spiralis'),\n",
    "        ('ustilago_maydis', 'Ustilago maydis'),\n",
    "        ('(ustilago)', 'Ustilago maydis'),\n",
    "        ('yarrowia_lipolytica', 'Yarrowia lipolytica'),\n",
    "        ('nasonia', 'Nasonia vitripennis'),\n",
    "        ('tomato', 'Solanum lycopersicum'),\n",
    "        ('chlamydomonas', 'Chlamydomonas reinhardtii'),\n",
    "        ('amphimedon', 'Amphimedon queenslandica'),\n",
    "        ('pneumocystis', 'Pneumocystis jirovecii'),\n",
    "        ('wheat', 'Triticum aestivum'),\n",
    "        ('chicken', 'Gallus gallus'),\n",
    "        ('zebrafish', 'Danio rerio'),\n",
    "        ('E_coli_K12', 'Escherichia coli'),\n",
    "        ('s_aureus', 'Staphylococcus aureus'),\n",
    "        ('volvox', 'Volvox carteri'),\n",
    "    ]\n",
    "]\n",
    "\n",
    "msg =\"Choose an identifier\"\n",
    "title = \"Identifier | Species\"\n",
    "chosen_species = eg.choicebox(msg, title, species_str)\n",
    "c_s = chosen_species.split(\":\")[1].replace(\" \",\"\")\n",
    "\n",
    "# Interactive window for file selection\n",
    "sequence = eg.fileopenbox(multiple=False)\n",
    "file_name = os.path.basename(sequence).split(\".\")[0]\n",
    "if not os.path.isdir(f\"{c_s}_{file_name}\"):\n",
    "    os.system(f\"mkdir {c_s}_{file_name}\") # Creating a folder for the given sequence\n",
    "os.chdir(f\"{c_s}_{file_name}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9affa48b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 01/12/2026 09:18:40\n",
      "New DB name:   /home/gman/Documents/MS-GENIOMHE/Projects/Anno2026/BruAnnoPipe/TE_DB/trep_db_renamed.fasta\n",
      "New DB title:  trep_db_renamed.fasta\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /home/gman/Documents/MS-GENIOMHE/Projects/Anno2026/BruAnnoPipe/TE_DB/trep_db_renamed.fasta\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 1866 sequences in 0.11331 seconds.\n",
      "\n",
      "\n",
      "Blast with TrepDB done.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "FASTA-Reader: Ignoring invalid residues at position(s): On line 106391: 3-4\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Softmasking + detection done.\n"
     ]
    }
   ],
   "source": [
    "# =========================\n",
    "#   TRANSPOSABLE ELEMENTS\n",
    "#   DETECTION AND SOFT MASKING \n",
    "# ========================\n",
    "if not os.path.isdir(f\"TE\"):\n",
    "    os.system(f\"mkdir TE\")\n",
    "os.chdir(\"../TE_DB\")\n",
    "\n",
    "# Using Blast with TrepDB to detect TE\n",
    "os.system(f\"makeblastdb -in trep_db_renamed.fasta -dbtype nucl -parse_seqids\")\n",
    "os.system(f\"blastn -db trep_db_renamed.fasta -query {sequence} -outfmt 7 -out ../{c_s}_{file_name}/TE/Blast_TE_output7\") \n",
    "os.system(f\"blastn -db trep_db_renamed.fasta -query {sequence} -out ../{c_s}_{file_name}/TE/Blast_TE_output\") \n",
    "print(\"Blast with TrepDB done.\")\n",
    "\n",
    "# Using Censor to detect and softmask TE\n",
    "os.system(\"echo 'GmandingoMAN' | sudo -S service tor stop\")\n",
    "os.system(\"echo 'GmandingoMAN' | sudo -S service tor start\")\n",
    "opts = Options()\n",
    "opts.headless = True\n",
    "opts.set_preference(\"print.print_headerleft\", \"\")\n",
    "opts.set_preference(\"print.print_headerright\", \"\")\n",
    "opts.set_preference(\"print.print_footerleft\", \"\")\n",
    "opts.set_preference(\"print.print_footerright\", \"\")\n",
    "\n",
    "output_pdf = os.path.join(os.path.abspath(f\"../{c_s}_{file_name}\"), \"result.pdf\")\n",
    "opts.set_preference(\"browser.helperApps.neverAsk.saveToDisk\", \"application/pdf\")\n",
    "opts.set_preference(\"print.always_print_silent\", True)\n",
    "opts.set_preference(\"print.print_to_file\", True)\n",
    "opts.set_preference(\"print.print_to_filename\", output_pdf)\n",
    "opts.set_preference(\"print_printer\", \"Mozilla Save to PDF\")\n",
    "opts.set_preference(\"print.printer_Mozilla_Save_to_PDF.print_to_file\", True)\n",
    "opts.set_preference(\"print.printer_Mozilla_Save_to_PDF.print_to_filename\", output_pdf)\n",
    "opts.set_preference(\"pdfjs.disabled\", True) \n",
    "\n",
    "# Basically I don't want to pay so I change my proxy with Tor (Kinda illegal btw)\n",
    "opts.set_preference(\"network.proxy.type\", 1)\n",
    "opts.set_preference(\"network.proxy.socks\", \"127.0.0.1\")\n",
    "opts.set_preference(\"network.proxy.socks_port\", 9050)\n",
    "opts.set_preference(\"network.proxy.socks_remote_dns\", True)\n",
    "\n",
    "driver = webdriver.Firefox(options=opts)\n",
    "driver.get(\"https://www.girinst.org/censor/\")\n",
    "driver.set_page_load_timeout(30)\n",
    "\n",
    "# Choosing the taxon\n",
    "name = chosen_species.split(\":\")[0][:-1]\n",
    "name1 = name\n",
    "if name == \"Triticum aestivum\":\n",
    "    name = \"Triticum\"\n",
    "driver.find_element(By.CSS_SELECTOR,\"select[name='taxon']\").click()\n",
    "options = driver.find_elements(By.CSS_SELECTOR, \"option\")\n",
    "for option in options:\n",
    "    if name in option.text:\n",
    "        option.click()\n",
    "driver.find_element(By.CSS_SELECTOR, \"div[id='content']\").click()\n",
    "\n",
    "# Okay time to submit the file\n",
    "file_input = driver.find_element(By.CSS_SELECTOR, \"input[type='file']\")\n",
    "file_input.send_keys(os.path.abspath(sequence))\n",
    "driver.find_element(By.CSS_SELECTOR, \"input[type='submit']\").click()\n",
    "\n",
    "# Waiting for the new page to load\n",
    "time.sleep(5)\n",
    "wait = WebDriverWait(driver, 40)\n",
    "link = wait.until(\n",
    "    EC.presence_of_element_located((By.XPATH, \"//a[@href]\"))\n",
    ")\n",
    "link.click()\n",
    "\n",
    "# Saving the results\n",
    "with open (f'../{c_s}_{file_name}/{file_name}_masked.fasta', 'w') as sq:\n",
    "    sq.write(driver.find_element(By.CSS_SELECTOR, \"pre\").text) # OMG ? The masked sequence is here\n",
    "driver.execute_script(\"window.print();\")\n",
    "os.chdir(f\"../{c_s}_{file_name}\")\n",
    "time.sleep(7)\n",
    "os.system(f\"mv result.pdf TE\")\n",
    "\n",
    "# Perfect we detected the TE !\n",
    "print(\"Softmasking + detection done.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7011617d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No prediction made by FGENESH\n",
      "Proteins and genes prediction done. Check the folders.\n"
     ]
    }
   ],
   "source": [
    "# =========================\n",
    "#   GENE PREDICTION\n",
    "# =========================\n",
    "def augustus(masked: bool):\n",
    "    # Using Augustus to predict genes\n",
    "    folder = \"augustus_pred\"\n",
    "    if masked :\n",
    "        os.system(f\"augustus --species={c_s} --codingseq=on --exonnames=on {file_name}_masked.fasta > augustus_pred_masked/augustus_result.txt\")\n",
    "        folder = \"augustus_pred_masked\"\n",
    "        os.system('grep -v \"^#\" augustus_pred_masked/augustus_result.txt > augustus_pred_masked/augustus_result.gff3')\n",
    "    else:\n",
    "        os.system(f\"augustus --species={c_s} --codingseq=on --exonnames=on {sequence} > augustus_pred/augustus_result.txt\")\n",
    "        os.system('grep -v \"^#\" augustus_pred/augustus_result.txt > augustus_pred/augustus_result.gff3')\n",
    "\n",
    "    # Extracting every predicted proteins, predicted CDS and predicted genes\n",
    "    with open(f\"{folder}/augustus_result.txt\", 'r', encoding='utf-8') as res:\n",
    "        with open(f\"{folder}/predicted_genes_by_augustus.txt\", 'w', encoding=\"utf-8\") as pred :\n",
    "            prots = \"\"\n",
    "            cds = \"\"\n",
    "            gene_found, protein_found, CDS_found = False, False, False\n",
    "            for line in res:\n",
    "                if line.startswith(\"# start gene\"):\n",
    "                    gene_found = True\n",
    "                if line.startswith(\"# end gene\"):\n",
    "                    pred.write(line)\n",
    "                    gene_found = False\n",
    "                if gene_found:\n",
    "                    if line.startswith(\"# coding sequence\"):\n",
    "                        CDS_found = True\n",
    "                    if line.startswith(\"# protein sequence\"):\n",
    "                        CDS_found = False\n",
    "                        protein_found = True\n",
    "                    if line.startswith(\"# Evidence \"):\n",
    "                        protein_found = False\n",
    "                    pred.write(line)\n",
    "                if CDS_found:\n",
    "                    line = line.replace(\"coding sequence = \", \"\").replace(\"# \", \"\")\n",
    "                    cds += line\n",
    "                if protein_found:\n",
    "                    line = line.replace(\"protein sequence = \", \"\").replace(\"# \", \"\")\n",
    "                    prots += line\n",
    "\n",
    "    # We'll use Blastp to check the protein predictions\n",
    "    if not os.path.isdir(f\"{folder}/predicted_proteins\"):\n",
    "        os.system(f\"mkdir {folder}/predicted_proteins\")\n",
    "\n",
    "    # Creating .fasta files for proteins\n",
    "    seqs = re.findall(r\"\\[(.*?)\\]\", prots, flags=re.DOTALL)\n",
    "    gene_number = 0\n",
    "    for seq in seqs : \n",
    "        gene_number += 1\n",
    "        with open(f\"{folder}/predicted_proteins/Protein{gene_number}.fasta\", \"w\", encoding='utf-8') as pro :\n",
    "            pro.write(f\">Protein{gene_number}\\n\")\n",
    "            pro.write(seq)\n",
    "\n",
    "    # We'll use Blastn to check the mRNA predictions\n",
    "    if not os.path.isdir(f\"{folder}/predicted_mRNA\"):\n",
    "        os.system(f\"mkdir {folder}/predicted_mRNA\")\n",
    "\n",
    "    # Creating .fasta files for mRNA\n",
    "    seqs1 = re.findall(r\"\\[(.*?)\\]\", cds, flags=re.DOTALL)\n",
    "    gene_number = 0\n",
    "    for seq in seqs1 : \n",
    "        gene_number += 1\n",
    "        with open(f\"{folder}/predicted_mRNA/mRNA{gene_number}.fasta\", \"w\", encoding='utf-8') as mrna :\n",
    "            mrna.write(f\">mRNA{gene_number}\\n\")\n",
    "            mrna.write(seq)\n",
    "\n",
    "def fgenesh(masked: bool):\n",
    "    driver.get(\"http://www.softberry.com/berry.phtml?topic=fgenesh&group=programs&subgroup=gfind\")\n",
    "    folder = \"FGENESH\"\n",
    "    if masked:\n",
    "        folder = \"FGENESH_masked\"\n",
    "\n",
    "    # Submitting the fasta file\n",
    "    file_input2 = driver.find_element(By.CSS_SELECTOR, \"input[type='file']\")\n",
    "    if masked :\n",
    "        file_input2.send_keys(os.path.abspath(f\"{file_name}_masked.fasta\"))\n",
    "    else :\n",
    "        file_input2.send_keys(os.path.abspath(f\"{sequence}\"))\n",
    "\n",
    "    try:\n",
    "        # Choosing the organism\n",
    "        driver.find_element(By.XPATH, \"//span[contains(text(), 'Organism')]\").click()\n",
    "        driver.find_element(By.CSS_SELECTOR, \"input[tabindex='5']\").send_keys(name)\n",
    "        driver.find_element(By.CSS_SELECTOR, \"li.active-result\").click() # First result\n",
    "\n",
    "        # Getting the .pdf file of the results\n",
    "        driver.find_element(By.CSS_SELECTOR, \"input[type='submit']\").click()\n",
    "        wait.until(\n",
    "            EC.presence_of_element_located((By.XPATH, \"//a[contains(text(), 'Show picture of predicted genes in PDF file')]\"))\n",
    "        ).click()\n",
    "        download_dir = \"/home/crakshay/Téléchargements\"\n",
    "        files = [os.path.join(download_dir, f) for f in os.listdir(download_dir)]\n",
    "        latest_pdf = max(files, key=os.path.getmtime)\n",
    "        os.system(f\"mv {latest_pdf} {folder}\")\n",
    "\n",
    "        # Parsing the .pdf file of the results\n",
    "        with open(f\"{folder}/getfile.pdf\", \"rb\") as f:\n",
    "            viewer = SimplePDFViewer(f)\n",
    "            doc = PDFDocument(f)\n",
    "            whole_file = \"\"\n",
    "            nb_pages = len(list(doc.pages()))+1\n",
    "            for page in range(1, nb_pages): \n",
    "                viewer.navigate(page)\n",
    "                viewer.render()\n",
    "                whole_file += \"\".join(viewer.canvas.strings) + \"\\n\"\n",
    "\n",
    "        # Creating .fasta files for proteins and mRNA\n",
    "        seqs = re.findall(r'>(FGENESH[^\\n>]*)([\\s\\S]*?)(?=>FGENESH|$)', whole_file, flags=re.DOTALL)\n",
    "        prot_number = 0\n",
    "        mRNA_number = 0\n",
    "        for seq in seqs : \n",
    "            if not 'mRNA' in seq[0]:\n",
    "                with open(f\"{folder}/predicted_proteins/FGENESH_prot{prot_number}.fasta\", \"w\", encoding='utf-8') as pro :\n",
    "                    prot_number += 1\n",
    "                    comma = seq[0].index(\",\")\n",
    "                    try:\n",
    "                        oof = seq[0].index(\"+\", comma)\n",
    "                    except:\n",
    "                        oof = seq[0].index(\"-\", comma)\n",
    "                    se = seq[0][:oof+1] + \"\\n\" + seq[0][oof+1:]\n",
    "                    pro.write(f\">{se+seq[1]}\")\n",
    "            else:\n",
    "                with open(f\"{folder}/predicted_mRNA/FGENESH_mRNA{mRNA_number}.fasta\", \"w\", encoding='utf-8') as mRNA :\n",
    "                    mRNA_number += 1\n",
    "                    comma = seq[0].index(\",\")\n",
    "                    try:\n",
    "                        oof = seq[0].index(\"+\", comma)\n",
    "                    except:\n",
    "                        oof = seq[0].index(\"-\", comma)\n",
    "                    se = seq[0][:oof+1] + \"\\n\" + seq[0][oof+1:]\n",
    "                    mRNA.write(f\">{se+seq[1]}\")\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "\n",
    "# Using Augustus to predict genes and proteins\n",
    "if not os.path.isdir(\"augustus_pred\"): # Creating a folder for the predictions\n",
    "    os.system(\"mkdir augustus_pred\")\n",
    "augustus(False)\n",
    "if not os.path.isdir(\"augustus_pred_masked\"): \n",
    "    os.system(\"mkdir augustus_pred_masked\")\n",
    "augustus(True)\n",
    "\n",
    "# Using FGENESH to predict genes and proteins\n",
    "use_fgenesh = eg.boolbox(\"Do you want to use FGENESH? (Plant Genome only !)\", choices=(\"YES\", \"NO\"))\n",
    "if use_fgenesh:\n",
    "    # Non softmasked\n",
    "    if not os.path.isdir(\"FGENESH\"):\n",
    "        os.system(\"mkdir FGENESH\")\n",
    "    if not os.path.isdir(\"FGENESH/predicted_proteins\"):\n",
    "        os.system(\"mkdir FGENESH/predicted_proteins\")\n",
    "    if not os.path.isdir(\"FGENESH/predicted_mRNA\"):\n",
    "        os.system(\"mkdir FGENESH/predicted_mRNA\")\n",
    "    fgenesh(False)\n",
    "\n",
    "    # Softmasked\n",
    "    if not os.path.isdir(\"FGENESH_masked\"):\n",
    "        os.system(\"mkdir FGENESH_masked\")\n",
    "    if not os.path.isdir(\"FGENESH_masked/predicted_proteins\"):\n",
    "        os.system(\"mkdir FGENESH_masked/predicted_proteins\")\n",
    "    if not os.path.isdir(\"FGENESH_masked/predicted_mRNA\"):\n",
    "        os.system(\"mkdir FGENESH_masked/predicted_mRNA\")\n",
    "    fgenesh(True)\n",
    "else :\n",
    "    print(\"No prediction made by FGENESH\")\n",
    "\n",
    "print(\"Proteins and genes prediction done. Check the folders.\")\n",
    "\n",
    "# Creating folders for the specific databases\n",
    "if not os.path.isdir(\"blastp_results\"): # Creating a folder for the predictions\n",
    "    os.system(\"mkdir blastp_results\")\n",
    "if not os.path.isdir(\"blastp_results/nr\"):\n",
    "    os.system(\"mkdir blastp_results/nr\")\n",
    "if not os.path.isdir(\"blastp_results/swiss\"):\n",
    "    os.system(\"mkdir blastp_results/swiss\")\n",
    "\n",
    "if not os.path.isdir(\"blastn_results\"): # Creating a folder for the mRNA predictions\n",
    "    os.system(\"mkdir blastn_results\")\n",
    "if not os.path.isdir(\"blastx_result\"): \n",
    "    # Creating a folder for overall protein sequences validation\n",
    "    os.system(\"mkdir blastx_result\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bd339876",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BlastP and BlastN done. Check the folders.\n",
      "BlastX done.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# =========================\n",
    "#   PROTEIN PREDICTIONS\n",
    "#   mRNA PREDICTIONS\n",
    "#   VALIDATION\n",
    "# =========================\n",
    "def blast(dr, fi, type, directory, BLOSUM, DATABASE):\n",
    "    try:\n",
    "        dr.get(f\"https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blast{type}&PAGE_TYPE=BlastSearch&LINK_LOC=blasthome\")\n",
    "    except TimeoutException:\n",
    "        dr.refresh()\n",
    "    file_input1 = dr.find_element(By.CSS_SELECTOR, \"input[type='file']\")\n",
    "    if type in \"px\":\n",
    "        if type == \"p\":\n",
    "            file_input1.send_keys(os.path.abspath(f\"{directory}/predicted_proteins/{fi}\"))\n",
    "        if type == \"x\":\n",
    "            file_input1.send_keys(os.path.abspath(fi))\n",
    "        dr.find_element(By.CSS_SELECTOR,\"select[name='DATABASE']\").click()\n",
    "        dr.find_element(By.CSS_SELECTOR, f\"option[value='{DATABASE}']\").click() # nr_cluster_seq / swissprot\n",
    "        dr.find_element(By.CSS_SELECTOR, \"button[class='usa-accordion-button']\").click()\n",
    "        dr.find_element(By.CSS_SELECTOR,\"select[name='MATRIX_NAME']\").click()\n",
    "        dr.find_element(By.CSS_SELECTOR, f\"option[value='{BLOSUM}']\").click() #BLOSUM62\n",
    "        dr.find_element(By.CSS_SELECTOR, \"input[class='blastbutton']\").click()\n",
    "    else :\n",
    "        # TSA = archive of computationally assembled transcript sequences from primary data such as ESTs and NGS.\n",
    "        file_input1.send_keys(os.path.abspath(f\"{directory}/predicted_mRNA/{fi}\"))\n",
    "        dr.find_element(By.CSS_SELECTOR,\"select[name='DATABASE']\").click()\n",
    "        dr.find_element(By.CSS_SELECTOR, f\"option[value='{DATABASE}']\").click() # TSA\n",
    "        dr.find_element(By.CSS_SELECTOR, \"input[id='qorganism']\").send_keys(name1)\n",
    "        first_li = wait.until(\n",
    "            EC.presence_of_element_located(\n",
    "                (By.CSS_SELECTOR, \"ul.ui-ncbiautocomplete-options li\") # First result\n",
    "            )\n",
    "        )\n",
    "        first_li.click()\n",
    "        dr.find_element(By.CSS_SELECTOR, \"input[class='blastbutton']\").click()\n",
    "\n",
    "def download(dr, file, destination, suffix):\n",
    "    wait.until(\n",
    "        EC.presence_of_element_located((By.XPATH, \"//h1[contains(text(), 'results for')]\"))\n",
    "    )\n",
    "    try :\n",
    "        dr.find_element(By.XPATH, \"//div[contains(text(), 'No significant similarity found')]\")\n",
    "    except:\n",
    "        time.sleep(3)\n",
    "        rid = dr.find_element(By.CSS_SELECTOR, \"a[title='BLAST search request ID']\").text\n",
    "        dr.set_page_load_timeout(5)\n",
    "        try:\n",
    "            dr.get(f\"https://blast.ncbi.nlm.nih.gov/Blast.cgi?RESULTS_FILE=on&RID={rid}&FORMAT_TYPE=SAM_SQ&FORMAT_OBJECT=Alignment& \\\n",
    "               DESCRIPTIONS=100&ALIGNMENTS=100&CMD=Get&DOWNLOAD_TEMPL=Results_All&ADV_VIEW=on\")\n",
    "        except:\n",
    "            download_dir = \"/home/gman/Downloads\"\n",
    "            files = [os.path.join(download_dir, f) for f in os.listdir(download_dir)]\n",
    "            latest_file = max(files, key=os.path.getmtime)\n",
    "            os.system(f\"mv {latest_file} {os.path.basename(file).split(\".\")[0]}{suffix}.txt\")\n",
    "            os.system(f\"mv {os.path.basename(file).split(\".\")[0]}{suffix}.txt {destination}\")\n",
    "            time.sleep(3)\n",
    "            dr.set_page_load_timeout(60)\n",
    "\n",
    "def validate_mRNA(file, directory, suffix):\n",
    "    # Blasting with TSA database\n",
    "    blast(driver, file, \"n\", directory, '', 'tsa_nt')\n",
    "    # Getting the results\n",
    "    download(driver, file, \"blastn_results\", suffix) \n",
    "\n",
    "def validate_proteins(file, directory, suffix):\n",
    "    # Blasting with NR database\n",
    "        # Blosum62\n",
    "    blast(driver, file, \"p\", directory, 'BLOSUM62', 'nr_cluster_seq')\n",
    "    driver.execute_script(\"window.open('');\")  \n",
    "    driver.switch_to.window(driver.window_handles[1])   \n",
    "\n",
    "    # Blasting with Swissprot database\n",
    "        # Blosum62\n",
    "    blast(driver, file, \"p\", directory, 'BLOSUM62', 'swissprot')\n",
    "    driver.execute_script(\"window.open('');\")  \n",
    "\n",
    "    # Getting the results\n",
    "    driver.switch_to.window(driver.window_handles[0])\n",
    "    download(driver, file, f\"blastp_results/nr\", suffix) # Blosum60\n",
    "    driver.switch_to.window(driver.window_handles[1])\n",
    "    download(driver, file, f\"blastp_results/swiss\", suffix) # Blosum60\n",
    "    driver.switch_to.window(driver.window_handles[0]) \n",
    "\n",
    "# Please Blast I need this... My protein is kinda homeless\n",
    "\n",
    "for file in os.listdir(\"augustus_pred/predicted_mRNA\"):\n",
    "    validate_mRNA(file, \"augustus_pred\", \"_ap\")\n",
    "for file in os.listdir(\"augustus_pred_masked/predicted_mRNA\"):\n",
    "    validate_mRNA(file, \"augustus_pred\", \"_apm\")\n",
    "\n",
    "for file in os.listdir(\"augustus_pred/predicted_proteins\"):\n",
    "    validate_proteins(file, \"augustus_pred\", \"_ap\")\n",
    "for file in os.listdir(\"augustus_pred_masked/predicted_proteins\"):\n",
    "    validate_proteins(file, \"augustus_pred\", \"_apm\")\n",
    "\n",
    "if use_fgenesh:\n",
    "    for file1 in os.listdir(\"FGENESH/predicted_mRNA\"):\n",
    "        validate_mRNA(file1, \"FGENESH\", \"_fg\")\n",
    "    for file1 in os.listdir(\"FGENESH_masked/predicted_mRNA\"):\n",
    "        validate_mRNA(file1, \"FGENESH_masked\", \"_fgm\")\n",
    "\n",
    "    for file1 in os.listdir(\"FGENESH/predicted_proteins\"):\n",
    "        validate_proteins(file1, \"FGENESH\", \"_fg\")\n",
    "    for file1 in os.listdir(\"FGENESH_masked/predicted_proteins\"):\n",
    "        validate_proteins(file1, \"FGENESH_masked\", \"_fgm\")\n",
    "\n",
    "print(\"BlastP and BlastN done. Check the folders.\")\n",
    "\n",
    "# Let's check if some proteins are missing...\n",
    "blast(driver, sequence, \"x\", \"\", 'BLOSUM62', 'swissprot')\n",
    "download(driver,sequence,\"blastx_result\", \"_\") \n",
    "print(\"BlastX done.\")\n",
    "\n",
    "# ARI ARI ARI ARIVEDERCCI\n",
    "driver.quit()\n",
    "os.system(\"sudo service tor stop\")\n",
    "\n",
    "# Little correction\n",
    "if use_fgenesh:\n",
    "    if not os.path.isdir(\"pred_by_fgenesh\"):\n",
    "        os.system(\"mkdir pred_by_fgenesh\")\n",
    "    os.system(\"mv FGENESH pred_by_fgenesh\")\n",
    "    os.system(\"mv FGENESH_masked pred_by_fgenesh\")\n",
    "\n",
    "if not os.path.isdir(\"pred_by_augustus\"):\n",
    "    os.system(\"mkdir pred_by_augustus\")\n",
    "os.system(\"mv augustus_pred pred_by_augustus\")\n",
    "os.system(\"mv augustus_pred_masked pred_by_augustus\")\n",
    "\n",
    "os.chdir(\"..\")\n",
    "if not os.path.isdir(\"Results\"):\n",
    "    os.system(\"mkdir Results\")\n",
    "os.system(f\"mv {c_s}_{file_name} Results\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
